### ISSUE TITLE
Flicking problem

### ISSUE THREAD

**Issue by caxieyou** [linebender/vello#334]
I have test a svg file, which is not that big, not as big as the CIA map case.

When I loaded the file, and zoom in, I find that the screen is flicking, some small parts not rendering correctly.

Is this because of float precision problem?

Quit sure there is no clipping in this file

![test](https://github.com/linebender/vello/assets/6970846/4335edb3-4c9b-4978-ba95-00fd594b4612)

**Comment by llsansun** [comment]
It is found that the config.n_drawobj in coarse.wgsl has a certain relationship. When it exceeds 65535, will there be conflicts between the data before 65535 and the data after 65535 due to the synchronization of the working group, resulting in problems with the graphics display. Is there a better solution to the 65535 drawing limit?hope it can be resolved.

**Comment by raphlinus** [comment]
Yes, a limit of 64k draw objects is a known problem, and has a straightforward solution. This issue can serve as the tracking bug for that. Thanks for the analysis!

**Comment by caxieyou** [comment]
can you tell which bug or issue link is there so we can know when it's fixed or has any progress?
thanks a lot

**Comment by caxieyou** [comment]
is this bug been fixed?

**Comment by raphlinus** [comment]
Not yet. The stroke rework is taking a lot longer than expected, though there is progress. This will be a high priority after that, and is also one of the items tracked in #302.

**Comment by DorianRudolph** [comment]
Is this the same issue?

https://github.com/linebender/vello/assets/10325838/161d7b60-ff85-488e-a042-470998ab17b3

**Comment by raphlinus** [comment]
No, that issue is caused by overflow of internal buffers (related to #366), which is in turn provoked by not culling lines and tiles that land outside the viewport. We do plan to work on all that.

**Comment by raphlinus** [comment]
I plan on addressing the 64k draw object problem shortly. There are three approaches that can be taken.

One is to conditionally apply a 3-level dispatch when the (workgroup size)^2 limit is crossed. This is what's done with pathtags, and I find it ugly. Among other things, it requires more permutations of shaders to be compiled, and there's also some complex conditional logic for which shaders to dispatch. I do have a local patch which is almost done, so it is perhaps the path of least resistance.

The second approach is inspired by a technique I saw in [FidelityFX sort](https://gpuopen.com/fidelityfx-parallel-sort/), and is implemented in my recent [sorting exploration](https://github.com/googlefonts/compute-shader-101/pull/31). In that approach, each workgroup iterates over `num_blocks_per_wg` blocks, where each block is the amount of data currently handled by a single workgroup (256 draw objects). In that way, the size of the sequence is not inherently bounded by workgroup sizes.

A drawback to the latter approach is that it may limit the amount of addressable parallelism. Doing a quick calculation, for very large inputs it will dispatch 64k threads, regardless of the size of the input. That is more threads than directly supported by any existing hardware (RTX 4090 has 16k), though it may limit opportunities for latency hiding.

An advantage to the latter approach is that it's two fewer dispatches.

As a future potential optimization, we may want to have more permutations (specialization by pipeline override) to (a) allow larger workgroups when the hardware supports it (the WebGPU spec only requires 256, which informs the choices we've made), and (b) support iteration over multiple elements per thread. The former is probably the best way to improve opportunities to exploit parallelism on powerful GPUs (1M threads should be plenty for at least a while) and has no real downside other than wiring up the plumbing. The latter is more of a tradeoff, as it improves bandwidth for large problems but limits parallelism for small ones. To switch between the two adaptively requires potentially compiling both variants (affecting cold-start time including shader compilation) and of course the complexity of the logic.

The third approach is to go back to single pass scan techniques, as was done in piet-gpu. We now know how to do this in WebGPU (see [Zulip thread](https://xi.zulipchat.com/#narrow/stream/197075-gpu/topic/Single.20pass.20prefix.20sum.20without.20barriers)) but the performance implications are mixed; in particular it would be a performance regression on Apple Silicon.

I'm most inclined to go with the second approach, as I think it's the best set of tradeoffs and admits additional optimization that would address the biggest shortcoming. I'll start on a PR, and if that goes well, probably apply the same technique to path tags.
